include/envoy/server/hot_restart.h

```diff
@@ -66,12 +64,21 @@ class HotRestart {
   virtual void shutdownParentAdmin(ShutdownParentAdminInfo& info) PURE;
 
   /**
-   * Tell our parent to gracefully terminate itself.
+   * Tell our parent process to gracefully terminate itself.
    */
   virtual void terminateParent() PURE;
 
   /**
-   * Shutdown the hot restarter.
+   * Merge stats_proto into stats_store, taking into account the stats values we've already
+   * seen transferred.
+   * @param stats_store the store whose stats will be updated.
+   * @param stats_proto the stats values we are updating with.
+   */
+  virtual void mergeParentStats(Stats::StoreRoot& stats_store,
```

Comment:
TIOLI, and I don't think I was being clear before, but personally I would not bother having this be part of the interface. I would just have this be a static helper function that independently tested and then called from server.cc stat flushing code after requesting parent stats like the other stat helper functions. I don't think having this as part of the interface adds much value.

Reply:
mergeParentStats needs to keep state from merge to merge. That state is held in the HotRestart machinery. Holding the state all the way out in Server::InstanceImpl would be unnecessary exposure of implementation details.

Reply:
What state does it need? AFAICT you just use what you pass in to the function?

Reply:
The state is the values that were previously passed into mergeParentStats, so that (new_parent - old_parent) can be added to the current stat value.

Reply:
I think we are talking past each other. My point is what you have now has nothing to do with hot restart. It's an independent class that stores its own state. Why do we need to call through the hot restarter interface to do this merging? It seems like something that can be done on the server side if we actually get stats from the parent? If you really prefer to keep it this way I will drop this point, but it seems like a bunch of interface indirection for no reason.

Reply:
Ok, now I see what you're saying, that Server::InstanceImpl could/should own the StatMerger. I see what you mean about reducing indirection, but I think it would introduce other complexity. First, InstanceImpl would now own one more thing. Second, in order to not keep the parent stat state around forever, the hot restarter drops that state upon the parent being terminated. There are a few code paths that do parent termination, and that dropping of state would have to be added to all of them.

In looking through the code again, I found some things to refactor, with the result that now more is contained within HotRestart. I think you might like the result: now Server::InstanceImpl just tells the restarter "take care of whatever parent merging might need to be done". So, less indirection in that rather than calling a passthrough, the thing passed through now decides to do that call on its own.